src/main/java/org/spongepowered/common/item/recipe/crafting/SpongeShapedCraftingRecipeBuilder.java

```diff
@@ -0,0 +1,151 @@
+        this.ingredientMap.put(symbol, ingredient);
+        return this;
+    }
+
+    @Override
+    public ShapedCraftingRecipe.Builder where(char symbol, @Nullable ItemStackSnapshot ingredient) throws IllegalArgumentException {
+        return where(symbol, (ItemStackSnapshot iss) -> match(ingredient, iss));
+    }
+
+    @Override
+    public ShapedCraftingRecipe.Builder result(ItemStackSnapshot result) {
+        this.result = result;
+        return this;
+    }
+
+    @Override
+    public ShapedCraftingRecipe build() {
+        checkState(!this.aisle.isEmpty(), "aisle has not been set");
+        checkState(!this.ingredientMap.isEmpty(), "no ingredients set");
+        checkState(this.result != null, "no result set");
+
+        net.minecraft.item.ItemStack stack = null;
+        Iterator<String> aisleIterator = this.aisle.iterator();
+        int width = aisleIterator.next().length();
+        int height = 1;
+
+        checkState(width > 0, "The aisle cannot be empty.");
+
+        while(aisleIterator.hasNext()) {
+            height++;
+
+            checkState(aisleIterator.next().length() == width,
+                    "The aisle has an inconsistent width.");
+        }
+
+        return new SpongeShapedCraftingRecipe(width, height, result, aisle, ingredientMap);
+    }
+
+    @Override
+    public ShapedCraftingRecipe.Builder from(ShapedCraftingRecipe value) {
+        this.aisle.clear();
+        this.aisle.addAll(value.getAisle());
+        this.ingredientMap = Maps.newHashMap(value.getIngredientPredicates());
+        this.result = value.getExemplaryResult();
+        return this;
+    }
+
+    @Override
+    public ShapedCraftingRecipe.Builder reset() {
+        this.aisle.clear();
+        this.ingredientMap.clear();
+        this.result = null;
+        return this;
+    }
+
+    /**
+     * Mimic the vanilla matching behavior,
+     * taken from {@link ShapedRecipes#checkMatch(InventoryCrafting, int, int, boolean)}
+     *
+     * @param recipeStack The stack required by the recipe
+     * @param inventoryStack The stack found in the inventory
+     * @return Whether the stacks match according to the vanilla Minecraft behavior
+     */
+    public static boolean match(ItemStackSnapshot recipeStack, ItemStackSnapshot inventoryStack) {
+        net.minecraft.item.ItemStack recipeNMS = ItemStackUtil.fromSnapshotToNative(recipeStack);
+        net.minecraft.item.ItemStack inventoryNMS = ItemStackUtil.fromSnapshotToNative(inventoryStack);
+
+        if (!recipeNMS.isEmpty() || !inventoryNMS.isEmpty())
+        {
+            if (recipeNMS.isEmpty() != inventoryNMS.isEmpty())
+            {
+                return false;
+            }
+
+            if (recipeNMS.getItem() != inventoryNMS.getItem())
+            {
+                return false;
+            }
+
+            if (recipeNMS.getMetadata() != 32767 && recipeNMS.getMetadata() != inventoryNMS.getMetadata())
```

Comment:
What is `32767` for a magic number here?
Either create a constant with a meaningful name or add a comment.